heu/library/algorithms/dgk/public_key.h

```diff
@@ -0,0 +1,105 @@
+// Copyright 2023 zhangwfjh
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "heu/library/algorithms/util/he_object.h"
+#include "heu/library/algorithms/util/mp_int.h"
+
+namespace heu::lib::algorithms::dgk {
+
+class PublicKey : public HeObject<PublicKey> {
+ public:
+  void Init(const MPInt& n, const MPInt& g, const MPInt& h, const MPInt& u);
+
+  const MPInt& N() const { return n_; }
+  const MPInt& G() const { return g_; }
+  const MPInt& H() const { return h_; }
+  const MPInt& U() const { return u_; }
+  MPInt PlainModule() const { return u_; }
+  MPInt PlaintextBound() const { return u_ / MPInt::_2_; }
+  MPInt CipherModule() const { return n_; }
+
+  bool operator==(const PublicKey&) const;
+  bool operator!=(const PublicKey&) const;
+  std::string ToString() const override;
+
+ public:
+  // Random element of form h^r mod n
+  MPInt RandomHr() const;
+  // Deterministic encryption
+  MPInt Encrypt(const MPInt&) const;
+  MPInt MapIntoMSpace(const MPInt&) const;
+  MPInt MapBackToZSpace(const MPInt&) const;
+
+  void MulMod(const MPInt& a, const MPInt& b, MPInt* dst) const {
+    lut_->m_space->MulMod(a, b, dst);
+  }
+
+ private:
+  MPInt n_, g_, h_, u_;
+
+  struct LUT {
+    std::unique_ptr<MontgomerySpace> m_space;  // m-space for mod n
+    std::unique_ptr<BaseTable> g_pow;          // powers of g mod n
+    std::unique_ptr<BaseTable> h_pow;          // powers of h mod n
+  };
```

Comment:
It is a bit strange，
Why are the member variables in LUT not declared as the original type, I mean, to remove the unique_ptr

Reply:
done